pkg/server/auth/jwt.go

```diff
@@ -0,0 +1,123 @@
+package auth
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"strings"
+
+	"github.com/coreos/go-oidc/v3/oidc"
+	"github.com/go-logr/logr"
+)
+
+// PrincipalGetter implementations are responsible for extracting a named
+// principal from an HTTP request.
+type PrincipalGetter interface {
+	// Principal extracts a principal from the http.Request.
+	// It's not an error for there to be no principal in the request.
+	Principal(r *http.Request) (*UserPrincipal, error)
+}
+
+// JWTCookiePrincipalGetter inspects a cookie for a JWT token
+// and returns a principal object.
+type JWTCookiePrincipalGetter struct {
+	log        logr.Logger
+	verifier   *oidc.IDTokenVerifier
+	cookieName string
+}
+
+func NewJWTCookiePrincipalGetter(log logr.Logger, verifier *oidc.IDTokenVerifier, cookieName string) PrincipalGetter {
```

Comment:
I am assuming our existing `jwt` machinery wasn't  quite right for this, hence the new implementation?

Reply:
Indeed, there is some overlap but they mostly serve different needs.
As I understand it, the existing `jwt` implementation is written to wrap/extend tokens that are issued from git providers and are used to interact with their APIs. 
This PR adds support for parsing ID tokens from OIDC issuers (which may or may not be git providers) in order to extract information about the logged-in user (email/groups) and make calls to the Kubernetes API (but not the git provider API).